gxtest/copyfilter.cpp

```diff
@@ -0,0 +1,562 @@
+  PixelFormat pixel_fmt;
+  GammaCorrection gamma;
+  u8 prev_copy_filter_sum;
+  u8 copy_filter_sum;
+  u8 next_copy_filter_sum;
+  bool intensity_fmt;
+};
+template <>
+struct fmt::formatter<CopyFilterTestContext>
+{
+  constexpr auto parse(format_parse_context& ctx) { return ctx.begin(); }
+  template <typename FormatContext>
+  auto format(const CopyFilterTestContext& test, FormatContext& ctx) const
+  {
+    return fmt::format_to(ctx.out(),
+                          "pixel_fmt: {}, gamma: {}, copy filter: {}/{}/{}, intensity: {}",
+                          test.pixel_fmt, test.gamma, test.prev_copy_filter_sum,
+                          test.copy_filter_sum, test.next_copy_filter_sum, test.intensity_fmt);
+  }
+};
+
+void SetCopyFilter(u8 prev_copy_filter_sum, u8 copy_filter_sum, u8 next_copy_filter_sum);
+
+GXTest::Vec4<u8> GenerateEFBColor(u16 x, u16 y)
+{
+  const u8 r = static_cast<u8>(x);
+  const u8 g = static_cast<u8>(y == 4 ? x : (y == 3 ? 255 : 0));
+  const u8 b = static_cast<u8>(x);
+  const u8 a = static_cast<u8>(x);
+  return {r, g, b, a};
+}
+
+static u32 GenerateEFBDepth(u16 x, u16 y)
+{
+  auto color = GenerateEFBColor(x, y);
+  return (u32(color.r) << 16) | (u32(color.g) << 8) | color.b;
+}
+
+static void SetPixelFormat(PixelFormat pixel_fmt)
+{
+  PEControl ctrl{.hex = BPMEM_ZCOMPARE << 24};
+  ctrl.pixel_format = pixel_fmt;
+  ctrl.zformat = DepthFormat::ZLINEAR;
+  ctrl.early_ztest = false;
+  CGX_LOAD_BP_REG(ctrl.hex);
+}
+
+static void FillEFB(PixelFormat pixel_fmt)
+{
+  // Don't set the format to Z24 here since we need RGB8 for our EFB copy+z-texture hack below
+  SetPixelFormat(pixel_fmt == PixelFormat::Z24 ? PixelFormat::RGB8_Z24 : pixel_fmt);
+
+  ZMode zmode{.hex = BPMEM_ZMODE << 24};
+  zmode.testenable = true;
+  zmode.func = CompareMode::Always;
+  zmode.updateenable = true;
+  CGX_LOAD_BP_REG(zmode.hex);
+
+  CGX_WaitForGpuToFinish();
+
+  CGX_PEPokeDither(false);
+  CGX_PEPokeAlphaUpdate(true);
+  CGX_PEPokeColorUpdate(true);
+  CGX_PEPokeBlendMode(GX_BM_NONE, SrcBlendFactor::Zero, DstBlendFactor::Zero, LogicOp::Set);
+  CGX_PEPokeAlphaRead(GX_READ_NONE);
+  CGX_PEPokeZMode(true, CompareMode::Always, true);
+
+  // For some reason GX_PokeARGB hangs when using this format
+  if (pixel_fmt == PixelFormat::RGB565_Z16)
+    return;
+
+  for (u16 x = 0; x < 256; x++)
+  {
+    for (u16 y = 0; y < 8; y++)
+    {
+      CGX_PokeARGB(x, y, GenerateEFBColor(x, y), pixel_fmt);
+      // GX_PokeZ doesn't seem to work at all
+      // CGX_PokeZ(x, y, GenerateEFBDepth(x, y), pixel_fmt);
+    }
+  }
+
+  if (pixel_fmt == PixelFormat::Z24)
+  {
+    // HACK: Since GX_PokeZ doesn't seem to work, we instead use an EFB copy and then
+    // draw over it using the z-texture feature to set the depth buffer.
+    SetCopyFilter(0, 64, 0);
+    // This value should be overridden, but it's recognizable if it shows up
+    CGX_LOAD_BP_REG(BPMEM_CLEAR_Z << 24 | 123456);
+    GXTest::CopyToTestBuffer(0, 0, 255, 7, {.clear = true});
```

Comment:
Will add `GX_InvalidateTexAll` here.  But relatedly, [`GCX_DoEfbCopyTex` uses `DCFlushRange`](https://github.com/dolphin-emu/hwtests/blob/2967cdcbe18f1fb35c1ff11c80d5593d99c3eff9/gxtest/cgx.cpp#L168), when it seems like it should use `DCInvalidateRange` - as far as I can tell that makes no difference in practice, but it seems dubious.

Reply:
Oh... that's a bug; Really should be invalidate. 

Since it flushes after submitting the EFB copy request, there is a theoretical race condition that the EFB copy could start before the flushing completes. But it would be near impossible to trigger, the write gather pipe would have to have the perfect alignment so that the last write triggers a flush, and even then the DCFlushRange would probably always be ahead of the EFB copy, unless an IRQ happened to come in and delay the flush. Do we even have IRQs enabled?

But it's risky enough that we should be fixing it, in another PR.

Reply:
#48